rllib/utils/replay_buffers/storage.py

```diff
@@ -0,0 +1,503 @@
+                buffer. After reaching this number, older samples will be
+                dropped to make space for new ones.
+            buffer_file: Optional buffer file to wite the data to. The file must not
+                exist and the file name must end with an `.dat` extension.
+        """
+        super().__init__(capacity)
+        self._buffer_file = buffer_file
+        self._rm_file_on_del = False
+        if not self._buffer_file:
+            self._rm_file_on_del = True
+            with NamedTemporaryFile(prefix='replay_buffer_', suffix='.dat') as f:
+                self._buffer_file = f.name
+        if os.path.exists(self._buffer_file):
+            raise ValueError("buffer_file must not exist: {}".format(self._buffer_file))
+        if not self._buffer_file.endswith(".dat"):
+            raise ValueError("buffer_file must end with '.dat' extension")
+        self._buffer_file = os.path.abspath(self._buffer_file)
+
+        # The actual storage (shelf / dict of SampleBatches).
+        if pickle.HIGHEST_PROTOCOL < 5:
+            logger.warning('Recommended pickle protocol is at least 5 '
+                           'for fast zero-copy access of arrays')
+        self._samples = shelve.open(
+            self._buffer_file[:-4],
+            flag='c',
+            protocol=pickle.HIGHEST_PROTOCOL
+        )
+        # Make sure shelve created correct file for storage
+        assert os.path.exists(self._buffer_file)
+
+    @ExperimentalAPI
+    @override(LocalStorage)
+    def get_state(self) -> Dict[str, Any]:
+        state = super().get_state()
+        state["_buffer_file"] = self._buffer_file
+        state["_rm_file_on_del"] = self._rm_file_on_del
+        state["_pkl_proto"] = pickle.HIGHEST_PROTOCOL
+        # Never delete file since it will be reused later
+        self._rm_file_on_del = False
+        return state
+
+    @ExperimentalAPI
+    @override(LocalStorage)
+    def set_state(self, state: Dict[str, Any]) -> None:
+        # Clean up existing storage
+        self._samples.close()
+        if self._buffer_file != state["_buffer_file"] and self._rm_file_on_del:
+            os.remove(self._buffer_file)
+        # Restore given storage
+        self._buffer_file = state["_buffer_file"]
+        self._rm_file_on_del = state["_rm_file_on_del"]
+        self._samples = shelve.open(
+            self._buffer_file[:-4],
+            flag='w',
+            protocol=state["_pkl_proto"]
+        )
+        super().set_state(state)
+
+    @override(LocalStorage)
+    def _get(self, i: int) -> SampleBatchType:
+        return self._samples[str(i)]
+
+    @override(LocalStorage)
+    def _set(self, idx: int, item: SampleBatchType) -> None:
+        self._warn_replay_capacity(item, self.capacity / item.count)
+        self._samples[str(idx)] = item
+        self._samples.sync()
+
+    @override(LocalStorage)
+    def _del(self, i: int) -> SampleBatchType:
+        # Do not delete item since this leads to continuously
```

Comment:
Would it be possible to raise an error here? I'd argue that this is not expected behaviour and could lead to errors further down the road.

Reply:
Indeed, I did not expect this behavior of `shelve`.  It forces us to skip the deletion of items since it will flood the disk otherwise. However, raising an error would break `OnDiskStorage`.

By the way, deleting items is just for lowering the memory footprint of Storage. I added this information to the docstring.

Reply:
@smorad I added the option of dynamic space allocation for OnDiskStorage to lower its consumption of disk space. Its footprint is now similar to InMemoryStorage. However, the mentioned alternatives are worth a try. I wonder how they compare regarding runtime.